Implement --remap-path-prefix

jsgf opened the pull request:
Remove experimental -Zremap-path-prefix-from/to, and replace it with
the stabilized --remap-path-prefix=from=to variant.

Implementation for #41555

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @petrochenkov (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

michaelwoerister commented on src/librustc/session/config.rs:
```diff
@@ -1968,6 +1948,18 @@ pub fn build_session_options_and_crate_config(matches: &getopts::Matches)
+        .into_iter()
+        .map(|remap| {
+            let mut parts = remap.rsplitn(2, '=');
+            match (parts.next(), parts.next()) {
```
The evaluation of components in tuple constructors is *probably* defined, but I still find this hard to read. And it's probably the wrong order because the iterator starts at the end `:)` Could you rewrite this to something like:
```rust
let mut parts = ...;
let to = parts.next();
let from = parts.next();
match (from, to) { ... }
```

jsgf replied:
Is there anything that implements `FromIterator<Item=T>` for `Option<(T,...)>`? I keep ending up in situations where it would be useful (mostly parsing with split), but I suspect it would clash with an existing `FromIterator` for `Option<T>` (unless specialization helps?).

michaelwoerister commented:
Thanks a lot, @jsgf! The implementation looks good to me, with the bug fixed.

I think we also need to update some documentation, right @rust-lang/docs? There's [an entry](https://doc.rust-lang.org/unstable-book/compiler-flags/remap-path-prefix.html) for the feature in the unstable book. Where should this be moved to?

jsgf commented:
Yes, documentation was my next question.

jsgf commented:
Removed unstable book documenation and updated rustc.1 man page, which seems to be the only place I could find written documentation on command-line options.

steveklabnik commented:
Yes, for now that's true. We haven't created a space for `rustc` docs yet; we wanted to.

michaelwoerister commented:
Alright then: @bors r+

Thanks for pushing this over the finish line, @jsgf!

@rust-lang/core, @rust-lang/compiler, @rust-lang/dev-tools: We are stabilizing path prefix remapping!

bors commented:
:pushpin: Commit ffd8ea7 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister ffd8ea72f545103070482da2bfcd6e4b7efe2e87 -->

jsgf commented:
Ah, I just found a problem - it looks like having this option on the command line changes the crate hash.

jsgf commented:
Looks like  #48019/#48162 is still an issue.

jsgf commented:
Rebasing to master seems to have fixed it, so maybe I just didn't have the other change before.

sanxiyn commented:
@bors r+

bors commented:
:pushpin: Commit 56a6828 has been approved by `sanxiyn`

<!-- @bors r=sanxiyn 56a68285332000c858e9aeba7d66a4ec66ebff91 -->
